dashboard/app/assets/javascripts/header.js.erb

```diff
@@ -2,110 +2,575 @@
+      var ajax = this.resolveAjax(args, dataType);
+      ajax.dataType = dataType;
+      ajax.success = fnSave || this.save;
+      ajax.error = this.onAjaxError;
+
+      return $.ajax(ajax);
+    }
+
+    // Load JSON data into this store.
+    UIStore.prototype.load = function(args) {
+      return this.$load(args, "json");
+    }
+
+    // Helper: Load XML and return it as a JSON structure
+    UIStore.prototype.loadXML = function(args) {
+      return this.$load(args, "xml");
+    }
+
+    // Helper: Load HTML into this store, as a string
+    UIStore.prototype.loadHTML = function(args) {
+      return this.$load(args, "html");
+    }
+
+    return UIStore;
+  })();
+
+
+
+
+  // class HeaderStore extends UIStore
+      /**
+     * @param data {{
+     *   title: string,
+     *   finishLink: string,
+     *   statsPath: string,
+     *   showStageLinks: boolean,
+     *   trophies: Object,
+     *   scriptId: number,
+     *   scriptLevelId: number,
+     *   currentLevelIndex: number,
+     *   levels: Array.<{
+     *     displayText: string,
+     *     status: string,
+     *     unplugged: boolean,
+     *     assessment: boolean
+     *   }>
+     * }}
+     */
+  HeaderStore = (function(_super) {
+    __extends(HeaderStore, _super);
+
+    // Must have a constructor
+    function HeaderStore() {
+      return HeaderStore.__super__.constructor.apply(this, arguments);
+    }
+
+    HeaderStore.prototype.resolveUrl = function(args, dataType) {
+      return "/popup/progress";
+    }
+
+    HeaderStore.prototype.resolveFilename = function(args, dataType) {
+      return "/progress_" + args.script_name + "_" + args.stage_id + "_" + args.level_id + "." + dataType;
+    }
+
+    return HeaderStore;
+  })(UIStore);
+
+
+
+  // ScriptStore: load the script progress (used in the header dropdown)
+  ScriptStore = (function(_super) {
+    __extends(ScriptStore, _super);
+
+    // Must have a constructor
+    function ScriptStore() {
+      return ScriptStore.__super__.constructor.apply(this, arguments);
+    }
+
+    ScriptStore.prototype.resolveUrl = function(args, dataType) {
+      return "/popup/stats";
+    }
+
+    ScriptStore.prototype.resolveFilename = function(args, dataType) {
+      return "/script_" + args.script_id + "." + dataType;
+    }
+
+    return ScriptStore;
+  })(UIStore);
+
+
+
+  // class HeaderProgress extends UIComponent
+  HeaderProgress = (function(_super) {
+    __extends(HeaderProgress, _super);
+
+    // Must have a constructor
+    function HeaderProgress($el, props) {
+      HeaderProgress.__super__.constructor.apply(this, arguments);
+
+      this.onModalClick = __bind(this.onModalClick, this);
+
+      return this;
+    }
+
+    HeaderProgress.prototype.componentDidMount = function() {
+      var _this = this;
+
+      // Create a new HeaderStore and bind it to my state.
+      // TODO: Is this the right place to put this?
```

Comment:
This is a fine place to put this for now, as we know that HeaderProgress is 1:1 with HeaderStore.  As we implement multiple components, we'll need a common place to instantiate singletons, and a way to pass those refs in.  Flux/React offer one way to do it.